src/lib/storage/src/lmdb.rs

```diff
@@ -159,6 +159,45 @@ impl LmdbBackend {
         .expect("Failed to run tokio task")
     }
 
+    pub async fn batch_upsert(
+        &self,
+        table: String,
+        data: Vec<(u128, Vec<u8>)>,
+    ) -> Result<(), StorageError> {
+        let env = self.env.clone();
+        tokio::task::spawn_blocking(move || {
+            let mut rw_txn = env.write_txn()?;
+
+            // Open or create the database for the given table
+            let db = env.create_database::<U128<BigEndian>, Bytes>(&mut rw_txn, Some(&table))?;
+
+            // Create a map of keys and their associated values
+            let keymap: HashMap<u128, &Vec<u8>> = data.iter().map(|(k, v)| (*k, v)).collect();
+
+            // Iterate through the keys in sorted order
+            let mut sorted_keys: Vec<u128> = keymap.keys().cloned().collect();
+            sorted_keys.sort();
+
+            // Iterate through the sorted keys to perform upserts
+            for key in sorted_keys {
+                // Check if the key already exists
+                if db.get(&rw_txn, &key)?.is_some() {
```

Comment:
Why check but do the same thing anyway? That's just an extra IO operation

Reply:
because if you dont then lmdb has to do it anyway and lmdb does it slower and per each into the whole database

Reply:
But you are doing the exact same function call, regardless of whether the condition fails or passes, so calling `get()` first won't make a difference.

Reply:
I tested